### PR TITLE
Refactor CMake code to link against gRPC.

### LINKED ISSUE #127: Document dependencies between submodules and cmake rules.
What submodule compiles what targets is a bit of a mess today.  We need to:

- document the current dependencies, and what submodules compile what targets.
- describe a way to discover new dependencies, as we add/remove/update submodules.

See discussion on #124 for context.

### PR DESCRIPTION
**Pull request by coryan**
In this change I describe a more principled approach to managing dependencies (such as gRPC, Protobuf, Abseil, cctz, or googletest) and implement that approach for gRPC.  I think this fixes #127.  It also makes it possible to compile gRPC as a submodule in Windows.

### CODE REVIEW

File: cmake/IncludeGrpc.cmake [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+    add_library(protobuf::libprotobuf ALIAS libprotobuf)
+
+    # The binary name is different on some platforms, use CMake magic to get it.
+    set(PROTOBUF_PROTOC_EXECUTABLE $<TARGET_FILE:protoc>)
+    mark_as_advanced(PROTOBUF_PROTOC_EXECUTABLE)
+    set(PROTOC_GRPCPP_PLUGIN_EXECUTABLE $<TARGET_FILE:grpc_cpp_plugin>)
+    mark_as_advanced(PROTOC_GRPCPP_PLUGIN_EXECUTABLE)
+elseif ("${GOOGLE_CLOUD_CPP_GRPC_PROVIDER}" STREQUAL "package"
+        OR "${GOOGLE_CLOUD_CPP_GRPC_PROVIDER}" STREQUAL "vcpkg")
+    find_package(GRPC REQUIRED grpc>=1.8)
```
Comment (hemant-qlogic) [inline cmake/IncludeGrpc.cmake]: Here Required grpc version is >=1.8 and in first CMakeList.txt file required grpc version is >=1.4
Same for protobuf version here >=3.5 and in first CMakeLists.txt file is > 3.0
Is it correct?

File: doc/managing-dependencies-in-cmake.md [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+   these targets using `find_package()`.  We will also use `find_package()` to
+   find the `protobuf` library.  The package introduces a
+   `protobuf::libprotobuf` target.
+
+1. [`pkg-config`](https://www.freedesktop.org/wiki/Software/pkg-config/) when
+   compiled and installed with GNU Make gRPC installs `pkg-config` support
+   files. When `GOOGLE_CLOUD_CPP_GRPC_PROVIDER` is set to `pkg-config` we will
+   define `INTERFACE` libraries for `gRPC::grpc++` and `gRPC::grpc`each `absl::*`
+   library used in `google-cloud-cpp`.  These `INTERFACE` libraries will set
+   their target [properties](https://cmake.org/cmake/help/v3.5/manual/cmake-properties.7.html#properties-on-targets)
```
Comment (hemant-qlogic) [inline doc/managing-dependencies-in-cmake.md]: URL Can be moved to new line as not fit in a single line.

File: cmake/IncludeGrpc.cmake (hunk 1 above)
Comment (coryan) [reply cmake/IncludeGrpc.cmake]: Yes.  gRPC does not compile with protobuf < 3.4.   I should actually update the gRPC version to 1.9, we ran into crashing bugs before that version.

File: doc/managing-dependencies-in-cmake.md (hunk 2 above)
Comment (coryan) [reply doc/managing-dependencies-in-cmake.md]: Use markdown reference links to make this more readable.

**Pull request merged**